src/mlpack/CMakeLists.txt

```diff
@@ -88,6 +88,35 @@ foreach(incl_file ${INCLUDE_FILES})
       BYPRODUCTS ${CMAKE_BINARY_DIR}/include/mlpack/${incl_file})
 endforeach()
 
+# In Case of R_Bindings Move all of these header and source files to
+# <package>/src/mlpack/ after the library is built.
+if (BUILD_R_BINDINGS)
+  # Collect all header and source files in the library.
+  file(GLOB_RECURSE R_SRC_HPP_FILES RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} *.hpp)
+  file(GLOB_RECURSE R_SRC_CPP_FILES RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} *.cpp)
+  set(R_SRC_FILES ${R_SRC_HPP_FILES} ${R_SRC_CPP_FILES})
+
+  # First we have to create that directory though.
+  execute_process(
+    COMMAND ${CMAKE_COMMAND} -E
+      make_directory ${CMAKE_BINARY_DIR}/src/mlpack/bindings/R/mlpack/src/mlpack
+  )
+  execute_process(
+    COMMAND ${CMAKE_COMMAND} -E
+      copy ${CMAKE_BINARY_DIR}/src/mlpack/mlpack_export.hpp
+           ${CMAKE_BINARY_DIR}/src/mlpack/bindings/R/mlpack/src/mlpack
+  )
+
+  # Then copy each of the header and source files over to that directory.
+  foreach(r_src_file ${R_SRC_FILES})
+    execute_process(
+      COMMAND ${CMAKE_COMMAND} -E copy_if_different
```

Comment:
Should we copy all the files here? As files present in julia, python, go, src/tests, bindings/tests folder might not required. Or should we provide complete mlpack with R package??

Reply:
I am sorry but is this the same question we have been discussing umpteen times?

Irrespective of the MLPACK repo, if you submit an *R* package to *CRAN* it has to build there.  And the lack of MLPACK *there* (and our hesitation of bundling "all of" MLPACK means there is no MLPACK at CRAN.

Or am I misunderstanding the question?

Reply:
Or maybe I completely misunderstood and you were just talking about moving from here (ie mlpack repo) to rcppmlpack (your other repo...)

Reply:
Right, definitely, understood.  I guess, the question is more along these lines---for Python and other languages, the typical process to make that available in that language's package manager / ecosystem is something that looks like a big long sandboxed build process that includes a step that does `make mlpack && make install` (thus making libmlpack.so and any other shared libraries used by the bindings), then, e.g., `make python` and then packaging up the generated files.  The binary package provided by that language's package manager, then, just includes any system-specific libraries that we built in that sandbox (through whatever means) and then the language-specific bindings themselves.  So, the mlpack PyPI package doesn't need any mlpack headers; it comes with libmlpack.so, a bunch of auto-generated `.so`s for each binding, an auto-generated `.h` file containing the symbols in those auto-generated `.so`s (for Cython), and then a bunch of .pyx files that use those shared objects.  (Same for Julia, roughly.)

However I don't know how R works.  Would those build processes be separate?  Could we use our existing CMake infrastructure during some "CRAN packaging job" to build `libmlpack.so` and all the other shared objects for each binding, like the other languages?  Or do we need to simply copy over all the hpp/cpp files that are a part of mlpack, and when a user calls a binding from R, CRAN/Rcpp(/whatever infrastructure I don't understand right now) will compile it on the spot using all the mlpack headers and sources that are available?  If it's the former, then we might not need these build steps.  If it's the latter, I suppose we do, but I don't understand enough about it.

Perhaps, the right way to handle this is to wait and see from my end as I learn more about how CRAN and R packaging works.

At the very least, I can say that we shouldn't need anything from `src/mlpack/bindings/[^R]/` or `src/mlpack/tests/`. :+1:

Reply:
This is not the best place to have these discussions.  But let's try.

So please step back.  

Consider an R package *without compiled code*.  It is self contained.  It needs build, and ideally run some tests (or examples). Have docs etc.

Now, consider an R package with a small amount of code it brings,  No external dependencies.  Basically the same thing again:  self contained. Same other requirements.

Now, consider an R package dependending on a system library.  Say libmysql.  _Either_ it runs some autoconf logic and expects it to be present and fails if not, _or_ it bundles it (if small enough),  (RMysql does the latter.)    

Sometimes we have hybrids.  Ie I helped nloptr to use a system libnlopt if one was present (to save time as the fallback was to go to MIT, download the nlopt source and build it; which sometimes fails and is always slower thna using an existing libnlopt).

NOW BACK TO US HERE.   We want a package, loosely named RcppMLPACK, to be _self-contained_ (as above in the other examples).   So if we put in an accessor function to call MLPACK's lars well .... then we also have to _somehow_ provide object code, likely from source or linking, and headers, to provide lars.   BUT THIS UP TO US.

Few CRAN packages build from a comparable large (and CMake driven) external repo.  So in a nutshell we need to copy whatever code we want used, ensure it gets build as a local *static library* that then gets linked into the RcppMLPACK.so (or .dylib) that R calls when the user says `library(RcppMLPACK)`.

I hope that made sense to you and was comprehensible.  If not, I am sorry but I tried.  There are many working examples to study at CRAN.

Reply:
Thanks for the explanation; it does make sense.  You're right that perhaps there could be a better place to have the discussion, but I think this works.  (Sorry Yashwant to hijack your thread a bit.)  To be honest, I think it would be easier in our situation if we *don't* use a hybrid package---i.e., we always provide bundled C++ sources for mlpack matching the exact version that was used to generate the bindings automatically.

(Perhaps we should split this following follow-up question into a separate issue; I can do that if you like, just let me know.)

I can see from the original RcppMLPACK sources that a [Makevars](https://github.com/rcppmlpack/RcppMLPACK1/blob/master/src/Makevars) file is used to build the library.  I don't know exactly how that is called but I assume that that's done during `library(RcppMLPACK)`.  My concern about this file is that it pretty clearly needs to be hand-maintained, but that sort of goes against the notion of "automatic" bindings.  The vision we should have is that when mlpack itself changes or a new binding is added or new sources are added or whatever, we can do `make R` without *any* further modification and what we get is something that can basically then be directly uploaded to CRAN, maybe with a few other automatable steps after `make R`.  So my questions are all from the angle of seeing how that is possible.

Thanks to your clarifications, I now have a better picture, allowing me to ask more questions. :smile:  So... when calling `library(RcppMLPACK)`, is CMake or anything like that available so that we can use mlpack's usual build process?  Ideally we would want to use that if we can.  (If it's not, I think that's okay, I'm sure we can come up with some way to automatically generate an appropriate `Makevars`.)

I hope I'm honing in on questions that indicate alignment. :)  I think my last question probably felt a little more out there.

Reply:
We are in the same (or adjacent) timezone(s); maybe us two should just get on the phone.

Having these discussions here makes no sense.  R packaging a booklength topic (if not more); I cannot answer that here is breadcrumb-sized units.   

I can't help if you jump to conclusion, but if you jump you still risk landing in the wrong spot. 

Look e.g. at the _automatically included `Makevars` for RcppArmadillo_ [here](https://github.com/RcppCore/RcppArmadillo/blob/master/inst/skeleton/Makevars).  It is so automated that it never needs modification (hence "automatic"?) _but it also takes advantage of a huge added benefit of being "in R":  LAPACK and BLAS given, no linking.

And that is out $64k issue.  MLPACK is different.  Cannot do without linking or embedding.  Yashwant addressed this so well already with a viable candiate package.

Reply:
I recently checked and about five packages (out of nearly 16k total of which nearly 4k have compiled code) use CMake.  We can use CMake if we want to, we simply have to declare it.

Reply:
Sure, I'm up for a call---I'll send an email.

Reply:
@Yashwants19 what did you end up thinking on the original question you asked here?  It seems like the code now doesn't copy any mlpack sources over to the R build directory.

Reply:
I have checked this is working fine, `methods` and `core` folder is copying properly. 
Is my approach is not correct here?

Reply:
Actually I'm not seeing the code that copies `methods/` and `core/`; maybe I just overlooked it.  Anyway, if it works, sounds good to me. :+1: